runtime-modules/project-token/src/lib.rs

```diff
@@ -237,40 +234,209 @@ decl_module! {
+        /// - some finished vesting schedule is removed from buyer's account_data in case
+        ///   number of buyer's vesting_schedules was == MaxVestingSchedulesPerAccountPerToken
+        ///   and a new schedule was added
+        /// - `token_data.last_sale.quantity_left` is decreased by `amount`
+        #[weight = 10_000_000] // TODO: adjust weight
+        pub fn purchase_tokens_on_sale(
+            origin,
+            token_id: T::TokenId,
+            amount: <T as Trait>::Balance
+        ) -> DispatchResult {
+            let current_block = Self::current_block();
+            let sender = ensure_signed(origin)?;
+            let token_data = Self::ensure_token_exists(token_id)?;
+            let sale = OfferingStateOf::<T>::ensure_sale_of::<T>(&token_data)?;
+            let sale_id = token_data.sales_initialized;
+            let joy_amount = sale.unit_price.saturating_mul(amount.into());
+            let account_exists = AccountInfoByTokenAndAccount::<T>::contains_key(token_id, &sender);
+            let bloat_bond = Self::bloat_bond();
+
+            let required_joy_balance = if account_exists {
+                joy_amount
+            } else {
+                joy_amount.saturating_add(bloat_bond)
+            };
+
+            // Ensure buyer's JOY balance is sufficient for the purchase
+            // and bloat bond (if required)
+            Self::ensure_can_transfer_joy(&sender, required_joy_balance)?;
+
+            // Ensure enough tokens are available on sale
+            ensure!(
+                sale.quantity_left >= amount,
+                Error::<T>::NotEnoughTokensOnSale
+            );
+
+            // Ensure participant's cap is not exceeded
+            if let Some(cap) = sale.cap_per_member {
+                Self::ensure_purchase_cap_not_exceeded(
+                    token_id,
+                    sale_id,
+                    &sender,
+                    amount,
+                    cap
+                )?;
+            }
+
+            // Ensure account exists if Permissioned token
+            if let TransferPolicy::Permissioned(_) = token_data.transfer_policy {
+                Self::ensure_account_data_exists(token_id, &sender)?;
+            }
+
+            // Ensure vesting schedule can added if doesn't already exist
+            // (MaxVestingSchedulesPerAccountPerToken not exceeded)
+            let acc_data = AccountInfoByTokenAndAccount::<T>::get(token_id, &sender);
+            let vesting_cleanup_key = acc_data.ensure_can_add_or_update_vesting_schedule::<T>(
+                current_block,
+                VestingSource::Sale(sale_id)
+            )?;
+
+            // == MUTATION SAFE ==
+
+            <Joy::<T> as Currency<T::AccountId>>::transfer(
+                &sender,
+                &sale.tokens_source,
+                joy_amount,
+                ExistenceRequirement::KeepAlive
+            )?;
+
+            if account_exists {
```

Comment:
So this code fragment was _super complicated_. It took me a very long time (>30m) to become comfortable with it actually doing what it should, but I am still not entirely sure. **I strongly encourage you to review one more time.** e.g. are we sure everyone gets a vesting schedule that starts at the same time (end of sale), not aligned with time of purchase?

The main points are how the crediting of the purchaser is happening (which is kind of hidden) and how vesting schedules are getting stacked with each purchase. I don't know if there is a better way to make this codepath simpler, but in either case, I think you should add a long meaty explainer section in the comments to just explain what is about to happen, as it will help the reader develop a mental model faster, which reduces risks of making mistakes on future changes.

Reply:
> are we sure everyone gets a vesting schedule that starts at the same time (end of sale), not aligned with time of purchase?

Yes. Every vesting schedule related to the sale is produced by `TokenSale::get_vesting_schedule`, which either:
- sets `VestingSchedule.start_block` to `self.start_block.saturating_add(self.duration)` (sale end block) or;
- sets `init_block` for `VestingSchedule::from_params` to `self.start_block.saturating_add(self.duration)` (sale end block), in which case `VestingSchedule.start_block = init_block.saturating_add(params.blocks_before_cliff)`, so `sale_end_block.saturating_add(params.blocks_before_cliff)`.

I think it will become much easier to parse if I just introduce `TokenSale::end_block()` method and rename `VestingSchedule::start_block` to something like `VestingSchedule::linear_vesting_start_block` or `VestingSchedule::cliff_block`, but I will add some more comments also.

Reply:
I added more comments and did some renaming which should make this logic clearer.
Done in https://github.com/Joystream/joystream/pull/3618/commits/d8e70c04967c6c8050a12b204be1f39974e955ef